pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,279 @@
+				for _, existingSubject := range existingBinding.Subjects {
+					if existingSubject.Kind == newSubject.Kind &&
+						existingSubject.Name == newSubject.Name &&
+						existingSubject.Namespace == newSubject.Namespace {
+						foundExisting = true
+						break
+					}
+				}
+				if !foundExisting {
+					newSubjects = append(newSubjects, newSubject)
+				}
+			}
+		} else {
+			newSubjects = binding.Subjects
+		}
+	} else {
+		return admission.NewForbidden(a,
+			fmt.Errorf("no policy binding found: %s", policy))
+	}
+
+	if len(newSubjects) == 0 {
+		return nil
+	}
+
+	matchersList, err := q.oclient.CollaboratorMatchers(ns).List(kapi.ListOptions{})
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	userLiterals := []string{}
+	userInGroupLiterals := []string{}
+	userSelectors := []labels.Selector{}
+	groupLiterals := []string{}
+	groupSelectors := []labels.Selector{}
+	serviceaccountLiterals := []string{}
+	serviceaccountInTheseNamespacesLiterals := []string{}
+
+	for _, matcher := range matchersList.Items {
+		// TODO: Use a cache for users, groups, and serviceaccounts.
+		switch {
+		case matcher.Spec.UserMatcher != nil:
+			userLiterals = append(userLiterals, matcher.Spec.UserMatcher.Users...)
+			userInGroupLiterals = append(userInGroupLiterals,
+				matcher.Spec.UserMatcher.UsersInTheseGroups...)
+			for _, labelSelector := range matcher.Spec.UserMatcher.UsersMatchingTheseSelectors {
+				selector, err := unversioned.LabelSelectorAsSelector(labelSelector)
+				if err != nil {
+					return admission.NewForbidden(a, err)
+				}
+				userSelectors = append(userSelectors, selector)
+			}
+		case matcher.Spec.GroupMatcher != nil:
+			groupLiterals = append(groupLiterals, matcher.Spec.GroupMatcher.Groups...)
+			for _, labelSelector := range matcher.Spec.GroupMatcher.GroupsMatchingTheseSelectors {
+				selector, err := unversioned.LabelSelectorAsSelector(labelSelector)
+				if err != nil {
+					return admission.NewForbidden(a, err)
+				}
+				groupSelectors = append(groupSelectors, selector)
+			}
+		case matcher.Spec.ServiceAccountMatcher != nil:
+			serviceaccountLiterals = append(serviceaccountLiterals,
+				matcher.Spec.ServiceAccountMatcher.ServiceAccounts...)
+			serviceaccountInTheseNamespacesLiterals =
+				append(serviceaccountInTheseNamespacesLiterals,
+					matcher.Spec.ServiceAccountMatcher.ServiceAccountsInTheseNamespaces...)
+		}
+	}
+
+	// TODO: Aggregate errors and return an error that lists all subjects that are
+	// not allowed.
+nextSubject:
+	for _, subj := range newSubjects {
+		switch subj.Kind {
+		case authorizationapi.SystemUserKind, authorizationapi.SystemGroupKind:
+		case authorizationapi.UserKind:
+			user, err := q.oclient.Users().Get(subj.Name)
+			if err != nil {
+				return err
+			}
+
+			for _, literal := range userLiterals {
+				if user.Name == literal {
+					continue nextSubject
+				}
+			}
+
+			for _, literal := range userInGroupLiterals {
+				for _, group := range user.Groups {
```

Comment:
There should be a groupcache which the API server already maintains to get group information for a user.  Use that instead of hitting the API for it.